### DISCUSSION TITLE
Hi，where the base function noreturn list is stored?

### DISCUSSION THREAD

**Discussion by baikaishiuc** [NationalSecurityAgency/ghidra#5608] [Q&A] [answered]
Hi, does Ghidra have a complete base list that indicates which functions are noreturn？

I know that there are many heuristics, or inference algorithms in ghidra that determine if a function is noreturn or not. but it is supposed to be deduced from a base table.

I searched the code and found that for example kernel32.hints has a noreturn list in it

```
<?xml version="1.0" encoding="UTF-8"?>
<LIBRARY_HINTS NAME="KERNEL32.DLL">
    <HINT ATTR="NO_RETURN" VALUE="y" NAME="ExitProcess" />
    <HINT ATTR="NO_RETURN" VALUE="y" NAME="ExitThread" />
    <HINT ATTR="NO_RETURN" VALUE="y" NAME="FreeLibraryAndExitThread" />
</LIBRARY_HINTS>

```

But this list is not enough, for example it doesn't contain _assert from UCRTBASE.dll, or some noreturn functions under linux, like exit.

Can you tell me where this list of basic noreturn functions is located?

**Comment by ryanmkurtz** [comment, marked as answer]
They are in `Ghidra/Features/Base/data/*FunctionsThatDoNotReturn`

**Reply by baikaishiuc** [reply]
cool, thanks

**Comment by baikaishiuc** [comment]
>
